src/VisualStudio/Core/Impl/SolutionExplorer/DiagnosticItem/DiagnosticItem.cs

```diff
@@ -12,81 +13,74 @@
 using Microsoft.Internal.VisualStudio.PlatformUI;
 using Microsoft.VisualStudio.Imaging;
 using Microsoft.VisualStudio.Imaging.Interop;
+using Roslyn.Utilities;
 
-namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer
-{
-    internal partial class DiagnosticItem : BaseItem
-    {
-        private readonly AnalyzerReference _analyzerReference;
-        private readonly IAnalyzersCommandHandler _commandHandler;
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.SolutionExplorer;
 
-        public ProjectId ProjectId { get; }
-        public DiagnosticDescriptor Descriptor { get; }
-        public ReportDiagnostic EffectiveSeverity { get; private set; }
-
-        public override event PropertyChangedEventHandler? PropertyChanged;
+internal sealed partial class DiagnosticItem(
+    ProjectId projectId,
+    AnalyzerReference analyzerReference,
+    DiagnosticDescriptor descriptor,
+    ReportDiagnostic effectiveSeverity,
+    IAnalyzersCommandHandler commandHandler)
+    : BaseItem(descriptor.Id + ": " + descriptor.Title), IEquatable<DiagnosticItem>
+{
+    private readonly AnalyzerReference _analyzerReference = analyzerReference;
+    private readonly IAnalyzersCommandHandler _commandHandler = commandHandler;
 
-        public DiagnosticItem(ProjectId projectId, AnalyzerReference analyzerReference, DiagnosticDescriptor descriptor, ReportDiagnostic effectiveSeverity, IAnalyzersCommandHandler commandHandler)
-            : base(descriptor.Id + ": " + descriptor.Title)
-        {
-            ProjectId = projectId;
-            _analyzerReference = analyzerReference;
-            Descriptor = descriptor;
-            EffectiveSeverity = effectiveSeverity;
-            _commandHandler = commandHandler;
-        }
+    public ProjectId ProjectId { get; } = projectId;
+    public DiagnosticDescriptor Descriptor { get; } = descriptor;
+    public ReportDiagnostic EffectiveSeverity { get; } = effectiveSeverity;
 
-        public override ImageMoniker IconMoniker
-            => MapEffectiveSeverityToIconMoniker(EffectiveSeverity);
+    public override ImageMoniker IconMoniker
+        => MapEffectiveSeverityToIconMoniker(EffectiveSeverity);
 
-        public override IContextMenuController ContextMenuController => _commandHandler.DiagnosticContextMenuController;
+    public override IContextMenuController ContextMenuController => _commandHandler.DiagnosticContextMenuController;
 
-        public override object GetBrowseObject()
-        {
-            return new BrowseObject(this);
-        }
+    public override object GetBrowseObject()
+        => new BrowseObject(this);
 
-        internal void UpdateEffectiveSeverity(ReportDiagnostic newEffectiveSeverity)
```

Comment:
we no longer update these in place (it makes equality/hash semantics non-ideal, which does impact how the observable collection and virtual tree view operate).